File: Python言語/Pythonで学ぶアルゴリズムの教科書 一生モノの知識と技術を身につける/chapter3_データ構造を学ぶ/3.リスト/list.py
# coding:utf-8
#   これを付けなければ、エラーが発生する(Python2だと判断された)。
print("朝来野智博")
print("リスト")

# 紐付き最大数を5つに制限している。
MAX = 5


def add_list(datalist, pointer, data):
    n = -1

    for ii in range(MAX):
        if datalist[ii] is None:
            n = ii
            break

    if n == -1:
        print("データ領域に空きがありません")

        return False, datalist, pointer

    for ii in range(MAX):
        if datalist[ii] is not None and pointer[ii] is None:
            pointer[ii] = n
            break

    datalist[n] = data
    pointer[n] = None
    print("データ", data, "追加。")

    return True, datalist, pointer


def del_list(datalist, pointer, head, data):
    n = -1

    for ii in range(MAX):
        if datalist[ii] == data:
            n = ii
            break

    if n == -1:
        print("データなし。")

        return False, datalist, pointer, head

    if n != head:
        for ii in range(MAX):
            if pointer[ii] == n:
                pointer[ii] = pointer[n]
    else:
        head = pointer[n]
        if head is None:
            head = 0
    datalist[n] = None
    pointer[n] = None
    print("データ", data, "削除。")

    return True, datalist, pointer, head


def put_list(datalist, pointer, head):
    pp = head

    while True:
        print(datalist[pp], end="--->")
        if pointer[pp] is None:
            print("EOF")
            break

        pp = pointer[pp]


def main():
    datalist = [None]*MAX
    pointer = [None]*MAX
    head = 0

    for ii in range(10, 70, 10):
        ret, datalist, pointer = add_list(datalist, pointer, ii)

    ret, datalist, pointer, head = del_list(datalist, pointer, head, 10)
    put_list(datalist, pointer, head)


main()
# 出力結果：
#           データ 10 追加。
#           データ 20 追加。
#           データ 30 追加。
#           データ 40 追加。
#           データ 50 追加。
#           データ領域に空きがありません
#           データ 10 削除。
#           20--->30--->40--->50--->EOF


print("以上。")
# 未だにPython2として認識される。
# vim: set requires python 3.9.7:
# vim: set ts=4 sts=4 sw=4 tw=0 ff=unix fenc=utf-8 ft=python expandtab:

File: Perl言語/Pythonで学ぶアルゴリズムの教科書 一生モノの知識と技術を身につける/chapter3_データ構造を学ぶ/3.リスト/list.pl
use v5.24;

my $asakuno = "朝来野智博";
my @asakuno = ("朝来野", "智博");

say "リスト";

sub add_list() {
	# データ投入。
	my ( $listdata, $ii ) = @_;

	# 以下、末尾に追加する。
	push @$listdata, $ii;
# 以下、先頭に追加する。
#	unshift @$listdata, $ii;
}

sub del_list() {
	# データ削除。
	my ( $listdata, $data) = @_;
	my @datalist = @$listdata;	# シャローコピー(浅い複製)とのこと。

	if ($#datalist <= 0) {
		return say "リストデータ存在せず。";
	}
	else {
		my $listdataIndex = -1;
#		while( my ($index, $value) = each @$listdata ){
		while( my ($index, $value) = each @datalist ){
			if ( "$value" eq "$data" ) {
				$listdataIndex = $index;
				last;	# シャローコピーとは言え、元のリファレンス配列から切り離せるようで、リファレンスカウントにはならずにループを抜け出せる。
			}
		}
		if ( $listdataIndex == -1 ) {
			return say "指定データ(" . $data . ")存在せず。";
		}
		else {
			# リストデータから削除。
			splice @$listdata, $listdataIndex, 1;
		}

		return @datalist;
	}
}

sub put_list() {
	my ($datalist) = @_;

	foreach my $value (@$datalist) {
		print "$value--->";
	}
	say "EOF";
}

sub main() {
	my @datalist;

	print "空表示：\n　　";
	&put_list(\@datalist);
	print "空削除：\n　　";
	&del_list(\@datalist, 0);

	for my $ii ( 0..3 ) {
		&add_list(\@datalist, $ii);		# データ追加(用意とも言う)。
	}
	&add_list(\@datalist, "asakuno");		# データ追加(用意とも言う)。
	for my $ii ( 4..5 ) {
		&add_list(\@datalist, $ii);
	}
	&add_list(\@datalist, "tomohiro");

	print "完全：\n　　";
	&put_list(\@datalist);

	print "3削除：\n　　";
	&del_list(\@datalist, 3);
	&put_list(\@datalist);

	print "asakuno削除：\n　　";
	&del_list(\@datalist, "asakuno");
	&put_list(\@datalist);

	print "先頭削除：\n　　";
	&del_list(\@datalist, 0);
	&put_list(\@datalist);
}
&main("asakuno");
# 出力結果：


say "以上。"
# vim: set ts=4 sts=4 sw=4 tw=0 ff=unix fenc=utf-8 ft=perl noexpandtab:
